### DISCUSSION TITLE
AWS S3 Storage Provider with Umbraco 12

### DISCUSSION THREAD

**Discussion by huzzi** [SixLabors/ImageSharp.Web#343] [Q&A] [answered]
I'm trying to use AWS S3 Storage Provider with Umbraco 12 but it's not working. The images are getting saved locally instead of AWS S3. 

This is my configuration code, have I missed anything?

```
            services.AddImageSharp().Configure<AWSS3StorageImageProviderOptions>(options =>
            {
                options.S3Buckets.Add(new AWSS3BucketClientOptions
                {
                    Endpoint = "{Endpoint}",
                    BucketName = "{BucketName}",
                    AccessKey = "{AccessKey}",
                    AccessSecret = "{AccessSecret}",
                    Region = "{Region}"
                });
            }).AddProvider<AWSS3StorageImageProvider>();`

```

**Comment by JimBobSquarePants** [comment]
Looks to me like you’ve only configured the provider not the cache.

https://docs.sixlabors.com/articles/imagesharp.web/gettingstarted.html

**Reply by huzzi** [reply]
I've configured the cache still not working.

**Reply by JimBobSquarePants** [reply]
Prove it

**Reply by huzzi** [reply]
This is my configuration
```

     public void ConfigureServices(IServiceCollection services)
        {
            services.AddUmbraco(_env, _config)
                .AddBackOffice()
                .AddWebsite()
                .AddComposers()
                .AddDependencyRegistration(_config)
                .Build();

            services.AddImageSharp().Configure<AWSS3StorageImageProviderOptions>(options =>
            {
                options.S3Buckets.Add(new AWSS3BucketClientOptions
                {
                    Endpoint = ",
                    BucketName = "",
                    AccessKey = "",
                    AccessSecret = "",
                    Region = "eu-west-1"
                });
            }).AddProvider<AWSS3StorageImageProvider>();

            services.AddImageSharp().Configure<AWSS3StorageCacheOptions>(options =>
                {
                    options.Endpoint = "";
                    options.BucketName = "";
                    options.AccessKey = "";
                    options.AccessSecret = ";
                    options.Region = "eu-west-1";
                    AWSS3StorageCache.CreateIfNotExists(options, S3CannedACL.Private);
                })
                .SetCache<AWSS3StorageCache>();
        }
```

 ```
 public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }
            app.UseImageSharp();

            app.UseUmbraco()
                .WithMiddleware(u =>
                {
                    u.UseBackOffice();
                    u.UseWebsite();
                })
                .WithEndpoints(u =>
                {
                    u.UseInstallerEndpoints();
                    u.UseBackOfficeEndpoints();
                    u.UseWebsiteEndpoints();
                });
        }
```
Is there anything I missed? Thanks

**Reply by JimBobSquarePants** [reply, marked as answer]
Why are you calling `services.AddImageSharp()` twice?

I would do 2 things.

1. compare your configuration to that in the unit test in this repository.
2. Ensure that Umbraco is not configuring the middleware. (Likely since they use it out of the box)

Tbh you should be asking all this in the Umbraco forums given that they customise configuration.

**Reply by ronaldbarendse** [reply]
I can confirm that Umbraco does indeed already call `AddImageSharp()` to setup the base configuration and calling it subsequently will overwrite it again with the defaults. The `IImageSharpBuilder` that's returned from this method is merely a wrapper with some nice helper extension methods, of which `IImageSharpBuilder.Configure<TOptions>()` simply forwards the call to `IServiceCollection.Configure<TOptions>()`...

It's also not clear to me whether you've already configured the Umbraco media file system to store the actual files in S3: at least your example code doesn't indicate that is the case. The [Umbraco.StorageProviders](https://github.com/umbraco/Umbraco.StorageProviders) project contains an Azure Blob Storage implementation and AWS S3 support is currently tracked in https://github.com/umbraco/Umbraco.StorageProviders/issues/28 (but the current implementations all rely on community efforts).
